scripts/httpd/upload.cgi

```diff
@@ -122,11 +122,54 @@ def ensure_namespaced(name, namespace):
     return name
 
 
-def main():
-    form = cgi.FieldStorage()
+def get_checksum_and_hash_type(form):
+    # Search for the file hash, start with stronger hash functions
+    if 'sha512sum' in form:
+        checksum = check_form(form, 'sha512sum')
+        hash_type = "sha512"
+
+    elif 'md5sum' in form:
+        # Fallback on md5, as it's what we currently use
+        checksum = check_form(form, 'md5sum')
+        hash_type = "md5"
+
+    else:
+        send_error('Required checksum is not present',
+                   status='400 Bad Request')
+
+    return checksum, hash_type
+
+
+def emit_fedmsg(config, name, checksum, filename, username, msgpath):
+    # Emit a fedmsg message.  Load the config to talk to the fedmsg-relay.
+    if config.getboolean('upload', 'fedmsgs', fallback=True):
+        import fedmsg
+        import fedmsg.config
```

Comment:
I think this will do the imports each time a message should be emitted? This doesn't seem to be ideal from the performance point of view. I am also thinking  what the exact behavior should be...either:
1) if fedmsg is configured and fedmsg is installed, then emit messages
2) if fedmsg is configured, try imports and fail if fedmsg is not installed

Seems like option 1) is better? I will need to give a bit more thought. What's your opinion?

Reply:
> This doesn't seem to be ideal from the performance point of view

True, it is not ideal because ideally, imports should be on top-level. But personally I don't think this can cause any performance issue. You do? I am not sure if I understand the architecture correctly (if not, please correct me), but I think we run this script on-demand and once it finishes its job, it exits, and then a new script is executed for a new upload. Is that right? In that case, we will import the package exactly the same times (once) because that function is called only once. And even if it wasn't the case, I hope that python is smart enough to know that the package is already imported and therefore do basically nothing when trying to import it again.

> if fedmsg is configured and fedmsg is installed, then emit messages
> if fedmsg is configured, try imports and fail if fedmsg is not installed

These sound good, I would like to implement them, but there is a problem. You have your `config` variable defined in `main()`. You cannot import the `fedmsg` package there because it won't be available in `emit_fedmsg` function. You don't want to move the `config` definition to top-level because then you would execute it even when `if __name__ == '__main__':` doesn't apply. You also don't want to move the fedmsg code back to the `main` function because ... because you want to do exactly the opposite - move everything possible from it to separate functions. It is already over-complicated.

I think those are all the considerable options. And from them, I think importing it in `emit_fedmsg` works best.

Or did I forget something?

Reply:
Right, it doesn't have any performance impact. I kind of didn't realize this is just a simple .cgi script.

I think we could move the import inside try, except block? And could we change the print in the except block to: `sys.stderr.write`. I know print was there before but I would like to change it as it is better not to bother user with a potential error there, only log it.

Reply:
Sure, sounds alright.

Reply:
Updated